### PR TITLE
Model builder: NVFP4 QMoE scheme and Model Optimizer checkpoint loader

### PR DESCRIPTION
**Pull request by tianleiwu**
### Description

Adds NVFP4 as a first-class model-builder quantization scheme and a loader for NVIDIA TensorRT Model Optimizer checkpoints.

* `quant_config.py` registers `nvfp4` as an `mx`-kind 4-bit dtype pinned to block size 16.
* `base.py` maps the MoE descriptor to the `QMoE` op's `quant_type`: `mxfp4` -> `"fp4"`, `nvfp4` -> `"nvfp4"`, integer dtypes -> `"int"`. Adds `make_fp8e4m3_initializer` for E4M3 block scales, plus `repack_modelopt_nvfp4_weight_codes` / `pack_nvfp4_codes_for_qmoe` to convert ModelOpt's `[N, K/2]` packed E2M1 layout into the layout `QMoE` expects.
* `builder.py` accepts `moe_quant_type=nvfp4` and validates it the same way as `mxfp4` (CUDA EP, symmetric INT4 precision, FP4 QMoE build).
* `quantized_model.py` gains `ModeloptModel` / `ModeloptDecoderLayer` (`quant_type="modelopt"`), which dequantizes ModelOpt NVFP4 (E2M1 codes + E4M3 block scale + fp32 global scale) and FP8 tensors so a ModelOpt checkpoint can be loaded like any other quantized checkpoint.

### Motivation and Context

Prerequisite for exporting NVFP4-quantized Qwen3.6 MoE checkpoints. On its own this only adds the scheme and the loader; the Qwen-specific export wiring is a follow-up.

Adds unit tests for the ModelOpt loader and for the NVFP4 repack round-trip (`test_modelopt_loader.py`, `test_nvfp4_repack.py`).

### CODE REVIEW

File: test/python/models/test_nvfp4_repack.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    packed_nk2 = ((high << 4) | low).astype(np.uint8)  # [N, K/2]
+
+    # Reference reconstruction from the modelopt-form tensors.
+    ref = np.empty((n, k), dtype=np.float32)
+    for i in range(n):
+        for j in range(k):
+            ref[i, j] = (
+                _decode_e2m1(codes_nk[i, j]) * _e4m3_byte_to_float(scale_bytes[i, j // block]) * float(global_scale)
+            )
+    return packed_nk2, scale_bytes, global_scale, ref
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply test/python/models/test_nvfp4_repack.py:76]: Fixed in 4cb54ae4. Both `_make_modelopt_nvfp4_projection`'s reference reconstruction and `_kernel_dequant_from_qmoe_layout` are now fully vectorized (NumPy nibble unpack + `_decode_e2m1` LUT gather + a single `torch.view(float8_e4m3fn)` decode of the whole block-scale array). The (2048, 512) case no longer runs per-element Python loops; the full file now runs in ~1s.

File: src/python/py/models/quantized_model.py [hunk 2]
```diff
@@ -1636,6 +1636,201 @@ def repack(self, module):
...
+    n = weight_u8.shape[0]
+    low = weight_u8 & 0x0F
+    high = weight_u8 >> 4
+    codes = torch.stack([low, high], dim=-1).reshape(n, -1).long()  # [N, K]
+    mag = _MODELOPT_FP4_E2M1_LUT[codes & 0x7]
+    val = torch.where((codes & 0x8) > 0, -mag, mag)  # [N, K]
+    bs = block_scale_e4m3.to(torch.float32)  # [N, K/16]
+    k = codes.shape[1]
+    bs = bs.repeat_interleave(k // bs.shape[1], dim=1)  # [N, K]
+    return (val * bs * float(global_scale)).to(torch.bfloat16)
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply src/python/py/models/quantized_model.py]: Fixed in 4cb54ae4. `_modelopt_dequant_nvfp4` now takes the tensor name and validates explicitly: missing `weight_scale` raises a named "has 'weight_scale_2' but no 'weight_scale'" error, raw `uint8` block scales are reinterpreted with `view(torch.float8_e4m3fn)` (not converted), any other dtype is rejected, and the block-scale shape is checked to be a clean block-wise split of the [N, K] weight. Added test coverage for the missing-scale and uint8-bytes cases.

File: src/python/py/models/quantized_model.py [hunk 3]
```diff
@@ -1636,6 +1636,201 @@ def repack(self, module):
...
+        for fname in files:
+            handle = self._open_handles.get(fname)
+            if handle is None:
+                handle = self._safe_open(os.path.join(self._input_path, fname), framework="pt", device="cpu")
+                self._open_handles[fname] = handle
+                self._handle_keys[fname] = set(handle.keys())
+            if name in self._handle_keys[fname]:
+                return handle.get_tensor(name)
+        return None
+
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply src/python/py/models/quantized_model.py:1814]: Fixed in 4cb54ae4. Added `close()` (which calls `__exit__` on each cached `safe_open` handle and clears the caches) plus a `__del__` safety net, and the loading body of `__init__` is now wrapped in `try/finally: self.close()`. Every tensor is materialized during `__init__`, so no descriptors stay open for the rest of the export; `_get` still re-opens lazily if called later. A test asserts `_open_handles` is empty after construction.

File: src/python/py/models/builder.py [hunk 4]
```diff
@@ -254,16 +254,19 @@ def check_extra_options(
...
                 )
             if not (precision == "int4" and extra_options.get("is_symmetric", True)):
                 raise ValueError(
-                    "moe_quant_type=mxfp4 requires building with precision=int4 (symmetric int4): the int4 build "
-                    "precision is what exports the quantized QMoE op, and mxfp4 only sets the MoE expert weights to "
-                    "the FP4 encoding."
+                    f"moe_quant_type={moe_quant_type} requires building with precision=int4 (symmetric int4): the "
+                    "int4 build precision is what exports the quantized QMoE op, and the FP4 scheme only sets the "
+                    "MoE expert weights to the FP4 encoding."
                 )
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply src/python/py/models/builder.py:270]: Good catch — fixed in 4cb54ae4. GPT-OSS ships MXFP4 experts, so there is no NVFP4 source to preserve and re-quantizing MXFP4 to NVFP4 is not implemented. `make_moe` now raises `moe_quant_type=nvfp4 is not supported for GPT-OSS. Use moe_quant_type=mxfp4.` before computing `is_fp4_moe`, instead of silently falling through to the integer QMoE path.

File: src/python/py/models/quantized_model.py [hunk 5]
```diff
@@ -1636,6 +1636,201 @@ def repack(self, module):
...
+        self.num_layers = n_layers
+
+        index_path = os.path.join(input_path, "model.safetensors.index.json")
+        if os.path.exists(index_path):
+            with open(index_path) as f:
+                self._weight_map = json.load(f)["weight_map"]
+            self._single_file = None
+        else:
+            self._weight_map = None
+            self._single_file = next(f for f in os.listdir(input_path) if f.endswith(".safetensors"))
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply src/python/py/models/quantized_model.py]: Fixed in 4cb54ae4. The no-index path now collects `sorted(...)` `.safetensors` candidates and raises a `ValueError` naming the directory and the actual count unless there is exactly one — no `StopIteration` and no arbitrary pick. Covered by a new test that deletes the index and adds a second shard.

File: src/python/py/models/quantized_model.py [hunk 6]
```diff
@@ -1636,6 +1636,201 @@
...
+        self.post_attention_layernorm = TensorModule()
+        self.self_attn = None
+        self.linear_attn = None
+        self.mlp = None
+
+    def is_empty(self):
+        return self.input_layernorm.weight is None
+
+
+class ModeloptModel(QuantizedModel):
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply src/python/py/models/quantized_model.py:1680]: Fixed in 4cb54ae4. `ModeloptModel` no longer derives from `QuantizedModel`. The base class eagerly loads every safetensors file into memory and unpacks/repacks integer quantized tensors, none of which applies to this loader (it streams tensors on demand and dequantizes to BF16), so calling `QuantizedModel.__init__` was never possible. The inheritance was only there for `modules()`, which is now a three-line method on the class; the builder consumes the loader duck-typed (`QuantModel.from_pretrained` has no `isinstance` contract).

**Review by tianleiwu — COMMENTED** [review]
Both suppressed comments from the latest review are addressed in b72fbb5e.

**1. `sys.modules` pollution in `test_nvfp4_repack.py`** — valid, and it was causing real failures. `_load_builder_model_class()` now records only the entries it actually inserts (an `inject()` helper that skips names already present, so a real `tqdm` imported by torch is never touched) and pops them in a `finally` block after `base.py` is loaded.

This was not hypothetical: running `pytest test/python/models test/python/builder` in one session, the leaked `MagicMock` broke a lazy import deep inside `torch.distributed.tensor`, which made `torch.testing.assert_close` blow up in `test_modelopt_loader.py`. Before/after on the same command: 53 failed / 370 passed -> 51 failed / 377 passed, with the failing-test set diff showing zero regressions and exactly those two loader tests fixed.

**2. `_build_layer()` silently assuming full attention** — also valid. The `else` branch is now an explicit `elif self._get(f"{p}.self_attn.q_proj.weight") is not None`, and anything else raises:

```
Layer {id} has neither '{p}.linear_attn.in_proj_qkv.weight' nor '{p}.self_attn.q_proj.weight',
so its attention variant cannot be determined. The Model Optimizer checkpoint is incomplete
or uses unsupported weight names.
```

Covered by a new case in `test_modelopt_loader_rejects_bad_checkpoints` that drops `self_attn.q_proj.weight` from a layer.

Note for a follow-up (out of scope here): several existing `test/python/builder/*` files stub `transformers` / `tqdm` / `onnxruntime` in `sys.modules` unconditionally and never restore them, so they have the same latent problem.

File: src/python/py/models/builders/base.py [hunk 7]
```diff
@@ -3890,6 +3896,52 @@ def make_fp8e8m0_initializer(self, scales_uint8, name):
...
+        from the raw uint8 bytes and tag it FLOAT8E4M3FN (the encoding the CUDA QMoE
+        NVFP4 kernel expects).
+        """
+        arr = scales_uint8.detach().cpu().numpy().astype(np.uint8) if isinstance(scales_uint8, torch.Tensor) else np.asarray(scales_uint8, dtype=np.uint8)
+        ir_tensor = ir.Tensor(np.ascontiguousarray(arr), dtype=ir.DataType.FLOAT8E4M3FN, name=name)
+        value = self.make_value(name, ir_tensor.dtype, ir_tensor.shape)
+        value.const_value = ir_tensor
+        self.model.graph.register_initializer(value)
+
+    @staticmethod
```
Comment (kunal-vaishnavi) [inline src/python/py/models/builders/base.py]: Can we remove any `staticmethod` decorators?

File: src/python/py/models/builders/base.py [hunk 8]
```diff
@@ -3890,6 +3896,52 @@ def make_fp8e8m0_initializer(self, scales_uint8, name):
...
+        if packed_nk2.dtype != torch.uint8:
+            packed_nk2 = packed_nk2.to(torch.uint8)
+        low = packed_nk2 & 0x0F
+        high = packed_nk2 >> 4
+        n = packed_nk2.shape[0]
+        codes = torch.stack((low, high), dim=-1).reshape(n, -1)  # [N, K]
+        return codes.contiguous()
+
+    @staticmethod
+    def pack_nvfp4_codes_for_qmoe(codes_nk):
```
Comment (kunal-vaishnavi) [inline src/python/py/models/builders/base.py]: Can we move the packing methods inside the `ModelOpt` class within `quantized_model.py`? Any unpacking and repacking for quantized models is already done there.

File: src/python/py/models/quantized_model.py [hunk 9]
```diff
@@ -1636,6 +1636,258 @@ def repack(self, module):
...
+# The ONNX Runtime GenAI QMoE op consumes the NVFP4 *routed* experts natively
+# (quant_type="nvfp4"), and the Qwen builder reads those per-expert tensors
+# straight from the source safetensors (see make_nvfp4_moe_initializers). So this
+# loader only has to materialize the NON-routed weights the builder walks as plain
+# modules: it dequantizes the FP8 attention and NVFP4 shared-expert / lm_head to
+# BF16 and passes BF16 tensors through unchanged. Attention is exported through the
+# builder's normal (int4 RTN) path since ONNX Runtime has no native FP8 attention
+# GEMM; dequantizing the FP8 weights to BF16 reconstructs them exactly.
+
+_MODELOPT_FP4_E2M1_LUT = torch.tensor([0.0, 0.5, 1.0, 1.5, 2.0, 3.0, 4.0, 6.0], dtype=torch.float32)
```
Comment (kunal-vaishnavi) [inline src/python/py/models/quantized_model.py]: All of these methods and variables that are prefixed with `_` can be inside the `ModelOpt` class.

File: src/python/py/models/quantized_model.py [hunk 10]
```diff
@@ -1636,6 +1636,258 @@ def repack(self, module):
...
+        self.mlp = None
+
+    def is_empty(self):
+        return self.input_layernorm.weight is None
+
+
+class ModeloptModel:
+    """Loader for NVIDIA Model Optimizer NVFP4 + FP8 mixed-precision checkpoints.
+
+    Deliberately not a ``QuantizedModel`` subclass: that base eagerly loads every
```
Comment (kunal-vaishnavi) [inline src/python/py/models/quantized_model.py]: Can we instead change the base class to do that? It would help with improving memory usage when unpacking/repacking beforehand vs on-the-fly.

File: src/python/py/models/builders/base.py (hunk 7 above)
Comment (tianleiwu) [reply src/python/py/models/builders/base.py]: Addressed in 1a4c3558b6. The NVFP4 packing helpers were removed from builders/base.py and are now ordinary instance methods on ModeloptModel, so the added staticmethod decorators are gone.

File: src/python/py/models/builders/base.py (hunk 8 above)
Comment (tianleiwu) [reply src/python/py/models/builders/base.py]: Addressed in 1a4c3558b6. repack_nvfp4_weight_codes() and pack_nvfp4_codes_for_qmoe() now live on ModeloptModel in quantized_model.py, and the repack test exercises them through that model instance.

File: src/python/py/models/quantized_model.py (hunk 9 above)
Comment (tianleiwu) [reply src/python/py/models/quantized_model.py]: Addressed in 1a4c3558b6. The ModelOpt FP4 lookup table and private dequantization helpers moved from module scope into ModeloptModel as class state and instance methods.

File: src/python/py/models/quantized_model.py (hunk 10 above)
Comment (tianleiwu) [reply src/python/py/models/quantized_model.py]: Addressed in 1a4c3558b6. ModeloptModel now derives from QuantizedModel, whose constructor supports load_weights=False to initialize the shared module surface without eagerly loading every shard. ModelOpt then streams tensors through _get() and materializes only the weights the builder needs.

File: src/python/py/models/quantized_model.py [hunk 11]
```diff
@@ -1636,6 +1641,258 @@
...
+        self._safe_open = safe_open
+        self._simple_namespace = SimpleNamespace
+        self._open_handles = {}
+        self._handle_keys = {}
+
+        with open(os.path.join(input_path, "config.json")) as f:
+            cfg = json.load(f)
+        tc = cfg.get("text_config", cfg)
+        n_layers = num_layers or tc.get("num_hidden_layers")
+        self.num_layers = n_layers
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply src/python/py/models/quantized_model.py]: Fixed in 967b829df2. ModeloptModel now resolves the config layer count before calling super() and passes the resolved value into QuantizedModel, so it no longer overwrites the inherited num_layers attribute.

File: src/python/py/models/quantized_model.py [hunk 12]
```diff
@@ -1636,6 +1641,258 @@
...
+            candidates = sorted(f for f in os.listdir(input_path) if f.endswith(".safetensors"))
+            if len(candidates) != 1:
+                raise ValueError(
+                    f"'{input_path}' has no 'model.safetensors.index.json', so it must contain exactly one "
+                    f".safetensors file, but found {len(candidates)}: {candidates}."
+                )
+            self._single_file = candidates[0]
+
+        try:
+            self.layers = [self._build_layer(layer_id) for layer_id in range(n_layers)]
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply src/python/py/models/quantized_model.py]: Fixed in 967b829df2. The non-eager base path initializes layers as a list, and ModeloptModel extends that inherited list in place instead of replacing it. The focused ModelOpt/NVFP4 suite passes (4 tests).

**Review by kunal-vaishnavi — APPROVED** [review]
